src/transcoding.ts

```diff
@@ -0,0 +1,318 @@
+      }
+      continue;
+    }
+
+    result[key] = request[key];
+  }
+
+  return result;
+}
+
+function camelToSnakeCase(str: string) {
+  return str.replace(/[A-Z]/g, letter => `_${letter.toLowerCase()}`);
+}
+
+export function requestCamelToSnakeCase(request: RequestType) {
+  if (!request || typeof request !== 'object') {
+    return request;
+  }
+  const convertedRequest: RequestType = {};
+  for (const field in request) {
+    const convertedField = camelToSnakeCase(field);
+    const value = request[field];
+    if (Array.isArray(value)) {
+      convertedRequest[convertedField] = value.map(v =>
+        requestCamelToSnakeCase(v as RequestType)
+      );
+    } else {
+      convertedRequest[convertedField] = requestCamelToSnakeCase(
+        value as RequestType
+      );
+    }
+  }
+  return convertedRequest;
+}
+
+export function transcode(
+  request: RequestType,
+  parsedOptions: ParsedOptionsType
+): TranscodedRequest | undefined {
+  const snakeRequest = requestCamelToSnakeCase(request);
+  const httpRules = [];
+  for (const option of parsedOptions) {
+    if (!(httpOptionName in option)) {
+      continue;
+    }
+
+    const httpRule = option[httpOptionName] as google.api.IHttpRule;
+    httpRules.push(httpRule);
+
+    if (httpRule?.additional_bindings) {
+      const additionalBindings = Array.isArray(httpRule.additional_bindings)
+        ? httpRule.additional_bindings
+        : [httpRule.additional_bindings];
+      httpRules.push(...additionalBindings);
+    }
+  }
+
+  for (const httpRule of httpRules) {
+    for (const httpMethod of supportedHttpMethods) {
+      if (!(httpMethod in httpRule)) {
+        continue;
+      }
+      const pathTemplate = httpRule[
+        httpMethod as keyof google.api.IHttpRule
+      ] as string;
+      const matchResult = match(snakeRequest, pathTemplate);
+      if (typeof matchResult === 'undefined') {
+        continue;
+      }
+      const {url, matchedFields} = matchResult;
+
+      if (httpRule.body === '*') {
+        // all fields except the matched fields go to request data
+        const data = deepCopy(snakeRequest);
+        for (const field of matchedFields) {
+          deleteField(data, field);
+        }
+        return {httpMethod, url, queryString: '', data};
+      }
+
+      // one field possibly goes to request data, others go to query string
+      const body = httpRule.body;
+      let data: string | RequestType = '';
+      const queryStringObject = deepCopy(snakeRequest);
+      if (body) {
+        deleteField(queryStringObject, body);
+        data = snakeRequest[body] as RequestType;
+      }
+      for (const field of matchedFields) {
+        deleteField(queryStringObject, field);
+      }
+      const queryStringComponents = buildQueryStringComponents(
+        queryStringObject
+      );
+      const queryString = queryStringComponents.join('&');
+      return {httpMethod, url, queryString, data};
+    }
+  }
+  return undefined;
+}
```

Comment:
Annoying ask - can we try to get 100% coverage on the added code?

Reply:
Working on it :)

Reply:
The best coverage will be when Showcase starts supporting REGAPIC (that's WIP by @vchudnov-g), but until then, I'll just try to cover as much as possible with unit tests.